affordance_primitives/test/test_task_monitor.cpp

```diff
@@ -0,0 +1,194 @@
+//          belonging to the University. None of the foregoing material may be
+//          copied or duplicated or disclosed without the express, written
+//          permission of the University. THE UNIVERSITY EXPRESSLY DISCLAIMS ANY
+//          AND ALL WARRANTIES CONCERNING THIS SOFTWARE AND DOCUMENTATION,
+//          INCLUDING ANY WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
+//          PARTICULAR PURPOSE, AND WARRANTIES OF PERFORMANCE, AND ANY WARRANTY
+//          THAT MIGHT OTHERWISE ARISE FROM COURSE OF DEALING OR USAGE OF TRADE.
+//          NO WARRANTY IS EITHER EXPRESS OR IMPLIED WITH RESPECT TO THE USE OF
+//          THE SOFTWARE OR DOCUMENTATION. Under no circumstances shall the
+//          University be liable for incidental, special, indirect, direct or
+//          consequential damages or loss of profits, interruption of business,
+//          or related expenses which may arise from use of software or
+//          documentation, including but not limited to those resulting from
+//          defects in software and/or documentation, or loss or inaccuracy of
+//          data of any kind.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include <gtest/gtest.h>
+#include <ros/ros.h>
+
+#include <affordance_primitives/task_monitor/task_monitor.hpp>
+#include <affordance_primitives/msg_types.hpp>
+
+#include <thread>
+
+// constexpr double EPSILON = 1e-4;
+
+using affordance_primitives::ExecutionResult;
+
+TEST(TaskMonitor, start_stop_monitor)
+{
+  ros::NodeHandle nh;
+
+  // Default parameters
+  affordance_primitives::AffordanceParameter params;
+
+  // Create and start monitor with no timeout
+  affordance_primitives::TaskMonitor monitor(nh, "ft_topic_name");
+  ASSERT_NO_THROW(monitor.startMonitor(params));
+
+  // Wait a sec, then stop monitor
+  ros::Duration(0.25).sleep();
+  ASSERT_NO_THROW(monitor.stopMonitor());
+
+  // Result should be STOP_REQUESTED
+  std::optional<ExecutionResult> result = monitor.getResult();
+  ASSERT_TRUE(result.has_value());
+  EXPECT_EQ(result.value(), affordance_primitives::ExecutionResult::STOP_REQUESTED);
+
+  // Now try interupting a timed run
+  ASSERT_NO_THROW(monitor.startMonitor(params, 0.25));
+  ASSERT_NO_THROW(monitor.stopMonitor());
+  result = monitor.getResult();
+  ASSERT_TRUE(result.has_value());
+  EXPECT_EQ(result.value(), affordance_primitives::ExecutionResult::STOP_REQUESTED);
+}
+
+TEST(TaskMonitor, monitor_timing)
+{
+  ros::NodeHandle nh;
+  const double test_duration = 0.25;
+
+  // Default parameters
+  affordance_primitives::AffordanceParameter params;
+
+  // Create and start monitor with no timeout
+  affordance_primitives::TaskMonitor monitor(nh, "ft_topic_name");
+  ASSERT_NO_THROW(monitor.startMonitor(params, test_duration));
+
+  // Immediately getting the result shouldn't work
+  EXPECT_FALSE(monitor.getResult().has_value());
+
+  // Wait some time and grab result
+  ros::Duration(1.05 * test_duration).sleep();
```

Comment:
Is there a way to programmatically determine wait time? I'm saying this because if for some reason there's a hangup on a resource that we're waiting for, won't this give false results?

Reply:
I set the task time in [line 76](https://github.com/UTNuclearRobotics/affordance_primitives/blob/c6ce220b19904d7d6d90e61c04029f7a6d787960/affordance_primitives/test/test_task_monitor.cpp#L76). I want it to be very short (tests should not take too long).

In this case, the `TaskMonitor` needs to wait through the `task_duration`, so this line (89) just waits 5% more than that to make sure the monitor ends